// twind/twind.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  __markAsModule(target);
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  if (module2 && module2.__esModule)
    return module2;
  return __exportStar(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", {value: module2, enumerable: true}), module2);
};

// src/index.ts
__export(exports, {
  autoprefix: () => autoprefix,
  create: () => create,
  cssomSheet: () => cssomSheet,
  hash: () => cyrb32,
  mode: () => mode,
  noprefix: () => noprefix,
  setup: () => setup,
  silent: () => silent,
  strict: () => strict,
  theme: () => theme2,
  tw: () => tw,
  voidSheet: () => voidSheet,
  warn: () => warn
});

// src/internal/is.ts
var string = (value) => typeof value === "string";
var object = (value) => value != null && typeof value === "object";
var fn = (value) => typeof value === "function";

// src/twind/util.ts
var includes = (value, search) => !!~value.indexOf(search);
var join = (parts, separator = "-") => parts.join(separator);
var joinTruthy = (parts, separator) => join(parts.filter(Boolean), separator);
var tail = (array, startIndex = 1) => array.slice(startIndex);
var identity = (value) => value;
var noop = () => {
};
var capitalize = (value) => value[0].toUpperCase() + tail(value);
var hyphenate = (value) => value.replace(/[A-Z]/g, "-$&").toLowerCase();
var escape = typeof CSS !== "undefined" && CSS.escape || ((className) => {
  const firstCodeUnit = className.charCodeAt(0);
  let firstChar = "";
  if (firstCodeUnit >= 48 && firstCodeUnit <= 57) {
    firstChar = "\\" + firstCodeUnit.toString(16) + " ";
    className = tail(className);
  }
  return firstChar + className.replace(/[!./:#]/g, "\\$&");
});
var cyrb32 = (value) => {
  let h = 9;
  for (let index = value.length; index--; ) {
    h = Math.imul(h ^ value.charCodeAt(index), 1597334677);
  }
  return "tw-" + ((h ^ h >>> 9) >>> 0).toString(36);
};
var sortedInsertionIndex = (array, element) => {
  let high = array.length;
  if (high === 0)
    return 0;
  for (let low = 0; low < high; ) {
    const pivot = high + low >> 1;
    if (array[pivot] <= element) {
      low = pivot + 1;
    } else {
      high = pivot;
    }
  }
  return high;
};

// src/twind/helpers.ts
var positions = (resolve) => (value, position2, prefix2, suffix) => {
  if (value) {
    const properties = position2 && resolve(position2);
    if (properties && properties.length > 0) {
      return properties.reduce((declarations, property2) => {
        declarations[joinTruthy([prefix2, property2, suffix])] = value;
        return declarations;
      }, {});
    }
  }
};
var corners = positions((key) => ({
  t: ["top-left", "top-right"],
  r: ["top-right", "bottom-right"],
  b: ["bottom-left", "bottom-right"],
  l: ["bottom-left", "top-left"],
  tl: ["top-left"],
  tr: ["top-right"],
  bl: ["bottom-left"],
  br: ["bottom-right"]
})[key]);
var expandEdges = (key) => {
  const parts = ({x: "lr", y: "tb"}[key] || key || "").split("").sort();
  for (let index = parts.length; index--; ) {
    if (!(parts[index] = {
      t: "top",
      r: "right",
      b: "bottom",
      l: "left"
    }[parts[index]]))
      return;
  }
  if (parts.length)
    return parts;
};
var edges = positions(expandEdges);

// src/twind/plugins.ts
var _;
var __;
var $;
var property = (property2) => (params, context, id) => ({
  [property2]: id + ((_ = join(params)) && "-" + _)
});
var propertyValue = (property2, separator) => (params) => ({
  [property2]: join(params, separator)
});
var themeProperty = (section) => (params, {theme: theme3}, id) => ({
  [section || id]: theme3(section || id, params)
});
var alias = (directive, name) => (params, context) => directive(params, context, name);
var display = property("display");
var position = property("position");
var textTransform = property("textTransform");
var textDecoration = property("textDecoration");
var fontStyle = property("fontStyle");
var fontVariantNumeric = (key) => (params, context, id) => ({
  ["--tw-" + key]: id,
  fontVariantNumeric: "var(--tw-ordinal,/*!*/ /*!*/) var(--tw-slashed-zero,/*!*/ /*!*/) var(--tw-numeric-figure,/*!*/ /*!*/) var(--tw-numeric-spacing,/*!*/ /*!*/) var(--tw-numeric-fraction,/*!*/ /*!*/)"
});
var inset = (params, {theme: theme3}, id) => ({[id]: theme3("inset", params)});
var opacityProperty = (params, theme3, id, section = id) => ({
  [`--tw-${id}-opacity`]: theme3(section + "Opacity", tail(params))
});
var parseColorComponent = (chars, factor) => Math.round(parseInt(chars, 16) * factor);
var asRGBA = (color, opacityProperty2, opacityDefault) => {
  if (color && color[0] === "#") {
    return `rgba(${parseColorComponent(color.substr(1, _ = (color.length - 1) / 3), $ = [17, 1, 0.062272][_ - 1])},${parseColorComponent(color.substr(1 + _, _), $)},${parseColorComponent(color.substr(1 + 2 * _, _), $)},var(--tw-${opacityProperty2}${opacityDefault ? "," + opacityDefault : ""}))`;
  }
  return color;
};
var withOpacityFallback = (property2, kind, color) => color && (_ = asRGBA(color, kind + "-opacity")) && _ !== color ? {
  [`--tw-${kind}-opacity`]: "1",
  [property2]: [color, _]
} : {[property2]: color};
var reversableEdge = (params, {theme: theme3}, id, section, prefix2, suffix) => (_ = {x: ["right", "left"], y: ["bottom", "top"]}[params[0]]) && ($ = `--tw-${id}-${params[0]}-reverse`) ? params[1] === "reverse" ? {
  [$]: "1"
} : {
  [$]: "0",
  [joinTruthy([prefix2, _[0], suffix])]: (__ = theme3(section, tail(params))) && `calc(${__} * var(${$}))`,
  [joinTruthy([prefix2, _[1], suffix])]: __ && [__, `calc(${__} * calc(1 - var(${$})))`]
} : void 0;
var placeHelper = (property2, params) => ({
  [property2]: (includes("wun", (params[0] || "")[3]) ? "space-" : "") + params[0]
});
var contentPluginFor = (property2) => (params) => includes(["start", "end"], params[0]) ? {[property2]: `flex-${params[0]}`} : placeHelper(property2, params);
var gridPlugin = (kind) => (params) => {
  switch (params[0]) {
    case "auto":
      return {[`grid-${kind}`]: "auto"};
    case "span":
      return {
        [`grid-${kind}`]: params[1] === "full" ? "1 / -1" : `span ${params[1]} / span ${params[1]}`
      };
    case "start":
    case "end":
      return {
        [`grid-${kind}-${params[0]}`]: params[1]
      };
  }
};
var border = (params, {theme: theme3}, id) => {
  switch (params[0]) {
    case "solid":
    case "dashed":
    case "dotted":
    case "double":
    case "none":
      return propertyValue("borderStyle")(params);
    case "collapse":
    case "separate":
      return propertyValue("borderCollapse")(params);
    case "opacity":
      return opacityProperty(params, theme3, id);
  }
  return (_ = theme3(id + "Width", params, "")) ? {borderWidth: _} : withOpacityFallback("borderColor", id, theme3(id + "Color", params));
};
var transform = (gpu) => (gpu ? "translate3d(var(--tw-translate-x,0),var(--tw-translate-y,0),0)" : "translateX(var(--tw-translate-x,0)) translateY(var(--tw-translate-y,0))") + " rotate(var(--tw-rotate,0)) skewX(var(--tw-skew-x,0)) skewY(var(--tw-skew-y,0)) scaleX(var(--tw-scale-x,1)) scaleY(var(--tw-scale-y,1))";
var transformXYFunction = (params, context, id) => (_ = context.theme(id, params[1] || params[0])) && {
  [`--tw-${id}-x`]: params[0] !== "y" && _,
  [`--tw-${id}-y`]: params[0] !== "x" && _,
  transform: [`${id}${params[1] ? params[0].toUpperCase() : ""}(${_})`, transform()]
};
var boxShadow = () => `var(--tw-ring-offset-shadow,0 0 transparent),var(--tw-ring-shadow,0 0 transparent),var(--tw-shadow,0 0 transparent)`;
var gradientColorStop = (params, {theme: theme3}, id) => (_ = theme3("gradientColorStops", params)) && {
  ["--tw-gradient-" + id]: _
};
var edgesPluginFor = (key) => (params, context, id) => id[1] ? edges(context.theme(key, params), id[1], key) : themeProperty(key)(params, context, id);
var padding = edgesPluginFor("padding");
var margin = edgesPluginFor("margin");
var minMax = (params, {theme: theme3}, id) => (_ = {w: "width", h: "height"}[params[0]]) && {
  [_ = `${id}${capitalize(_)}`]: theme3(_, tail(params))
};
var corePlugins = {
  group: (params, {tag}) => tag("group"),
  hidden: alias(display, "none"),
  inline: display,
  block: display,
  contents: display,
  flow: display,
  table: (params, context, id) => includes(["auto", "fixed"], params[0]) ? {tableLayout: params[0]} : display(params, context, id),
  flex(params, context, id) {
    switch (params[0]) {
      case "row":
      case "col":
        return {
          flexDirection: join(params[0] === "col" ? ["column", ...tail(params)] : params)
        };
      case "nowrap":
      case "wrap":
        return {flexWrap: join(params)};
      case "grow":
      case "shrink":
        return {[`flex-${params[0]}`]: params[1] || "1"};
    }
    return (_ = context.theme("flex", params, "")) ? {flex: _} : display(params, context, id);
  },
  grid(params, context, id) {
    switch (params[0]) {
      case "cols":
      case "rows":
        return params.length > 1 && {
          [`grid-template-${params[0] === "cols" ? "columns" : params[0]}`]: params.length === 2 && Number(params[1]) ? `repeat(${params[1]},minmax(0,1fr))` : join(tail(params), " ")
        };
      case "flow":
        return params.length > 1 && {
          gridAutoFlow: join(params[1] === "col" ? ["column", ...tail(params, 2)] : tail(params), " ")
        };
    }
    return display(params, context, id);
  },
  auto: (params) => includes(["cols", "rows"], params[0]) && (_ = params.length === 2 ? {
    auto: "auto",
    min: "min-content",
    max: "max-content",
    fr: "minmax(0,1fr)"
  }[params[1]] || `minmax(0,${params[1]})` : params.length > 2 && `minmax(${join(tail(params), ",")})`) && {
    [`grid-auto-${params[0] === "cols" ? "columns" : "rows"}`]: _
  },
  static: position,
  fixed: position,
  absolute: position,
  relative: position,
  sticky: position,
  visible: {visibility: "visible"},
  invisible: {visibility: "hidden"},
  antialiased: {
    WebkitFontSmoothing: "antialiased",
    MozOsxFontSmoothing: "grayscale"
  },
  "subpixel-antialiased": {
    WebkitFontSmoothing: "auto",
    MozOsxFontSmoothing: "auto"
  },
  truncate: {
    overflow: "hidden",
    whiteSpace: "nowrap",
    textOverflow: "ellipsis"
  },
  "sr-only": {
    position: "absolute",
    width: "1px",
    height: "1px",
    padding: "0",
    margin: "-1px",
    overflow: "hidden",
    whiteSpace: "nowrap",
    clip: "rect(0,0,0,0)",
    borderWidth: "0"
  },
  "not-sr-only": {
    position: "static",
    width: "auto",
    height: "auto",
    padding: "0",
    margin: "0",
    overflow: "visible",
    whiteSpace: "normal",
    clip: "auto"
  },
  resize: (params) => ({
    resize: {x: "vertical", y: "horizontal"}[params[0]] || params[0] || "both"
  }),
  box: (params) => ({"box-sizing": `${params[0]}-box`}),
  appearance: propertyValue("appearance"),
  cursor: propertyValue("cursor"),
  float: propertyValue("float"),
  clear: propertyValue("clear"),
  top: inset,
  right: inset,
  bottom: inset,
  left: inset,
  inset: (params, {theme: theme3}) => (_ = expandEdges(params[0])) ? edges(theme3("inset", tail(params)), params[0]) : (_ = theme3("inset", params)) && {
    top: _,
    right: _,
    bottom: _,
    left: _
  },
  underline: textDecoration,
  "line-through": textDecoration,
  "no-underline": alias(textDecoration, "none"),
  "text-underline": alias(textDecoration, "underline"),
  "text-no-underline": alias(textDecoration, "none"),
  "text-line-through": alias(textDecoration, "line-through"),
  uppercase: textTransform,
  lowercase: textTransform,
  capitalize: textTransform,
  "normal-case": alias(textTransform, "none"),
  "text-normal-case": alias(textTransform, "none"),
  italic: fontStyle,
  "not-italic": alias(fontStyle, "normal"),
  "font-italic": alias(fontStyle, "italic"),
  "font-not-italic": alias(fontStyle, "normal"),
  font: (params, context, id) => (_ = context.theme("fontFamily", params, "")) ? {fontFamily: _} : themeProperty("fontWeight")(params, context, id),
  items: (params) => ({
    alignItems: includes(["start", "end"], params[0]) ? `flex-${params[0]}` : join(params)
  }),
  content: contentPluginFor("alignContent"),
  justify: contentPluginFor("justifyContent"),
  self: contentPluginFor("alignSelf"),
  place: (params) => placeHelper("place-" + params[0], tail(params)),
  overscroll: (params) => ({
    ["overscrollBehavior" + (params[1] ? "-" + params[0] : "")]: params[1] || params[0]
  }),
  col: gridPlugin("column"),
  row: gridPlugin("row"),
  duration: themeProperty("transitionDuration"),
  delay: themeProperty("transitionDelay"),
  tracking: themeProperty("letterSpacing"),
  leading: themeProperty("lineHeight"),
  z: themeProperty("zIndex"),
  opacity: themeProperty(),
  ease: themeProperty("transitionTimingFunction"),
  p: padding,
  py: padding,
  px: padding,
  pt: padding,
  pr: padding,
  pb: padding,
  pl: padding,
  m: margin,
  my: margin,
  mx: margin,
  mt: margin,
  mr: margin,
  mb: margin,
  ml: margin,
  w: themeProperty("width"),
  h: themeProperty("height"),
  min: minMax,
  max: minMax,
  fill: themeProperty(),
  order: themeProperty(),
  origin: propertyValue("transformOrigin", " "),
  select: propertyValue("userSelect"),
  "pointer-events": propertyValue("pointerEvents"),
  align: propertyValue("verticalAlign"),
  whitespace: propertyValue("whiteSpace"),
  "normal-nums": {fontVariantNumeric: "normal"},
  ordinal: fontVariantNumeric("ordinal"),
  "slashed-zero": fontVariantNumeric("slashed-zero"),
  "lining-nums": fontVariantNumeric("numeric-figure"),
  "oldstyle-nums": fontVariantNumeric("numeric-figure"),
  "proportional-nums": fontVariantNumeric("numeric-spacing"),
  "tabular-nums": fontVariantNumeric("numeric-spacing"),
  "diagonal-fractions": fontVariantNumeric("numeric-fraction"),
  "stacked-fractions": fontVariantNumeric("numeric-fraction"),
  overflow: (params) => includes(["ellipsis", "clip"], params[0]) ? propertyValue("textOverflow")(params) : params[1] ? {[`overflow-${params[0]}`]: params[1]} : propertyValue("overflow")(params),
  transform: (params) => params[0] === "none" ? {transform: "none"} : {
    "--tw-translate-x": "0",
    "--tw-translate-y": "0",
    "--tw-rotate": "0",
    "--tw-skew-x": "0",
    "--tw-skew-y": "0",
    "--tw-scale-x": "1",
    "--tw-scale-y": "1",
    transform: transform(params[0] === "gpu")
  },
  rotate: (params, {theme: theme3}) => (_ = theme3("rotate", params)) && {
    "--tw-rotate": _,
    transform: [`rotate(${_})`, transform()]
  },
  scale: transformXYFunction,
  translate: transformXYFunction,
  skew: transformXYFunction,
  gap: (params, context, id) => (_ = {x: "column", y: "row"}[params[0]]) ? {[_ + "Gap"]: context.theme("gap", tail(params))} : themeProperty("gap")(params, context, id),
  stroke: (params, context, id) => (_ = context.theme("stroke", params, "")) ? {stroke: _} : themeProperty("strokeWidth")(params, context, id),
  outline: (params, {theme: theme3}) => (_ = theme3("outline", params)) && {
    outline: _[0],
    outlineOffset: _[1]
  },
  break(params) {
    switch (params[0]) {
      case "normal":
        return {
          wordBreak: "normal",
          overflowWrap: "normal"
        };
      case "words":
        return {overflowWrap: "break-word"};
      case "all":
        return {wordBreak: "break-all"};
    }
  },
  text(params, {theme: theme3}, id) {
    switch (params[0]) {
      case "left":
      case "center":
      case "right":
      case "justify":
        return {textAlign: params[0]};
      case "uppercase":
      case "lowercase":
      case "capitalize":
        return textTransform([], _, params[0]);
      case "opacity":
        return opacityProperty(params, theme3, id);
    }
    const fontSize = theme3("fontSize", params, "");
    if (fontSize) {
      return string(fontSize) ? {fontSize} : {
        fontSize: fontSize[0],
        ...string(fontSize[1]) ? {lineHeight: fontSize[1]} : fontSize[1]
      };
    }
    return withOpacityFallback("color", "text", theme3("textColor", params));
  },
  bg(params, {theme: theme3}, id) {
    switch (params[0]) {
      case "fixed":
      case "local":
      case "scroll":
        return propertyValue("backgroundAttachment", ",")(params);
      case "bottom":
      case "center":
      case "left":
      case "right":
      case "top":
        return propertyValue("backgroundPosition", " ")(params);
      case "no":
        return params[1] === "repeat" && propertyValue("backgroundRepeat")(params);
      case "auto":
      case "cover":
      case "contain":
        return propertyValue("backgroundSize")(params);
      case "repeat":
        return includes("xy", params[1]) ? propertyValue("backgroundRepeat")(params) : {"background-repeat": params[1] || params[0]};
      case "opacity":
        return opacityProperty(params, theme3, id, "background");
      case "clip":
        return {backgroundClip: params[1] + (params[1] === "text" ? "" : "-box")};
      case "gradient":
        if (params[1] === "to" && (_ = expandEdges(params[2]))) {
          return {
            backgroundImage: `linear-gradient(to ${join(_, " ")},var(--tw-gradient-stops,var(--tw-gradient-from,transparent),var(--tw-gradient-to,transparent)))`
          };
        }
    }
    return (_ = theme3("backgroundImage", params, "")) ? {backgroundImage: _} : withOpacityFallback("backgroundColor", "bg", theme3("backgroundColor", params));
  },
  from: gradientColorStop,
  to: gradientColorStop,
  via: (params, {theme: theme3}) => (_ = theme3("gradientColorStops", params)) && {
    "--tw-gradient-stops": `var(--tw-gradient-from,transparent),${_},var(--tw-gradient-to,transparent)`
  },
  border: (params, context, id) => expandEdges(params[0]) ? edges(context.theme("borderWidth", tail(params)), params[0], "border", "width") : border(params, context, id),
  divide: (params, context, id) => (_ = reversableEdge(params, context, id, "divideWidth", "border", "width") || border(params, context, id)) && {
    "&>:not([hidden])~:not([hidden])": _
  },
  space: (params, context, id) => (_ = reversableEdge(params, context, id, "space", "margin")) && {
    "&>:not([hidden])~:not([hidden])": _
  },
  placeholder: (params, {theme: theme3}, id) => (_ = params[0] === "opacity" ? opacityProperty(params, theme3, id) : withOpacityFallback("color", "placeholder", theme3("placeholderColor", params))) && {
    "&::placeholder": _
  },
  shadow: (params, {theme: theme3}) => (_ = theme3("boxShadow", params)) && {
    "--tw-shadow": _,
    boxShadow: [_, boxShadow()]
  },
  animate: (params, {theme: theme3, tag}) => {
    if ($ = theme3("animation", params)) {
      const parts = $.split(" ");
      if ((_ = theme3("keyframes", parts[0], __ = {})) !== __) {
        return ($ = tag(parts[0])) && {
          animation: $ + " " + join(tail(parts), " "),
          ["@keyframes " + $]: _
        };
      }
      return {animation: $};
    }
  },
  ring(params, {theme: theme3}, id) {
    switch (params[0]) {
      case "inset":
        return {"--tw-ring-inset": "inset"};
      case "opacity":
        return opacityProperty(params, theme3, id);
      case "offset":
        return (_ = theme3("ringOffsetWidth", tail(params), "")) ? {
          "--tw-ring-offset-width": _
        } : {
          "--tw-ring-offset-color": theme3("ringOffsetColor", tail(params))
        };
    }
    return (_ = theme3("ringWidth", params, "")) ? {
      "--tw-ring-offset-shadow": `var(--tw-ring-inset,/*!*/ /*!*/) 0 0 0 var(--tw-ring-offset-width,${theme3("ringOffsetWidth", "", "0px")}) var(--tw-ring-offset-color,${theme3("ringOffsetColor", "", "#fff")})`,
      "--tw-ring-shadow": `var(--tw-ring-inset,/*!*/ /*!*/) 0 0 0 calc(${_} + var(--tw-ring-offset-width,${theme3("ringOffsetWidth", "", "0px")})) var(--tw-ring-color,${asRGBA(theme3("ringColor", "", "#93c5fd"), "ring-opacity", theme3("ringOpacity", "", "0.5"))})`,
      "box-shadow": boxShadow()
    } : {
      "--tw-ring-opacity": "1",
      "--tw-ring-color": asRGBA(theme3("ringColor", params), "ring-opacity")
    };
  },
  object: (params) => includes(["contain", "cover", "fill", "none", "scale"], params[0]) ? propertyValue("objectFit")(params) : propertyValue("objectPosition", " ")(params),
  list: (params) => propertyValue(includes(["inside", "outside"], params[0]) ? "listStylePosition" : "listStyleType")(params),
  rounded: (params, context, id) => corners(context.theme("borderRadius", tail(params), ""), params[0], "border", "radius") || themeProperty("borderRadius")(params, context, id),
  "transition-none": {"transition-property": "none"},
  transition: (params, {theme: theme3}) => ({
    transition: joinTruthy([
      theme3("transitionProperty", params),
      theme3("transitionDuration", ""),
      theme3("transitionTimingFunction", "")
    ], " ")
  }),
  container: (params, {theme: theme3}) => {
    const {screens = theme3("screens"), center, padding: padding2} = theme3("container");
    const paddingFor = (screen) => (_ = padding2 && (string(padding2) ? padding2 : padding2[screen] || padding2.DEFAULT)) ? {
      paddingRight: _,
      paddingLeft: _
    } : {};
    return Object.keys(screens).reduce((rules2, screen) => {
      if (_ = screens[screen]) {
        rules2[`@media (min-width: ${_})`] = {
          "&": {
            "max-width": _,
            ...paddingFor(screen)
          }
        };
      }
      return rules2;
    }, {
      width: "100%",
      ...center ? {marginRight: "auto", marginLeft: "auto"} : {},
      ...paddingFor("xs")
    });
  }
};

// src/twind/preflight.ts
var createPreflight = (theme3) => ({
  ":root": {tabSize: 4},
  "body,blockquote,dl,dd,h1,h2,h3,h4,h5,h6,hr,figure,p,pre,fieldset,ol,ul": {margin: "0"},
  button: {backgroundColor: "transparent", backgroundImage: "none"},
  'button,[type="button"],[type="reset"],[type="submit"]': {WebkitAppearance: "button"},
  "button:focus": {outline: ["1px dotted", "5px auto -webkit-focus-ring-color"]},
  "fieldset,ol,ul,legend": {padding: "0"},
  "ol,ul": {listStyle: "none"},
  html: {
    lineHeight: "1.5",
    WebkitTextSizeAdjust: "100%",
    fontFamily: theme3("fontFamily", "sans", "ui-sans-serif,system-ui,sans-serif")
  },
  body: {fontFamily: "inherit", lineHeight: "inherit"},
  "*,::before,::after": {
    boxSizing: "border-box",
    border: `0 solid ${theme3("borderColor", "DEFAULT", "currentColor")}`
  },
  hr: {height: "0", color: "inherit", borderTopWidth: "1px"},
  img: {borderStyle: "solid"},
  textarea: {resize: "vertical"},
  "input::placeholder,textarea::placeholder": {
    color: theme3("placeholderColor", "DEFAULT", "#a1a1aa")
  },
  'button,[role="button"]': {cursor: "pointer"},
  table: {textIndent: "0", borderColor: "inherit", borderCollapse: "collapse"},
  "h1,h2,h3,h4,h5,h6": {fontSize: "inherit", fontWeight: "inherit"},
  a: {color: "inherit", textDecoration: "inherit"},
  "button,input,optgroup,select,textarea": {
    fontFamily: "inherit",
    fontSize: "100%",
    margin: "0",
    padding: "0",
    lineHeight: "inherit",
    color: "inherit"
  },
  "button,select": {textTransform: "none"},
  "::-moz-focus-inner": {borderStyle: "none", padding: "0"},
  ":-moz-focusring": {outline: "1px dotted ButtonText"},
  ":-moz-ui-invalid": {boxShadow: "none"},
  progress: {verticalAlign: "baseline"},
  "::-webkit-inner-spin-button,::-webkit-outer-spin-button": {height: "auto"},
  '[type="search"]': {WebkitAppearance: "textfield", outlineOffset: "-2px"},
  "::-webkit-search-decoration": {WebkitAppearance: "none"},
  "::-webkit-file-upload-button": {WebkitAppearance: "button", font: "inherit"},
  summary: {display: "list-item"},
  "abbr[title]": {textDecoration: "underline dotted"},
  "b,strong": {fontWeight: "bolder"},
  "pre,code,kbd,samp": {
    fontFamily: theme3("fontFamily", "mono", "ui-monospace,monospace"),
    fontSize: "1em"
  },
  "sub,sup": {fontSize: "75%", lineHeight: "0", position: "relative", verticalAlign: "baseline"},
  sub: {bottom: "-0.25em"},
  sup: {top: "-0.5em"},
  "img,svg,video,canvas,audio,iframe,embed,object": {display: "block", verticalAlign: "middle"},
  "img,video": {maxWidth: "100%", height: "auto"}
});

// src/twind/variants.ts
var coreVariants = {
  dark: "@media (prefers-color-scheme:dark)",
  sticky: "@supports ((position: -webkit-sticky) or (position:sticky))",
  "motion-reduce": "@media (prefers-reduced-motion:reduce)",
  "motion-safe": "@media (prefers-reduced-motion:no-preference)",
  first: "&:first-child",
  last: "&:last-child",
  even: "&:nth-child(2n)",
  odd: "&:nth-child(odd)",
  children: "&>*",
  siblings: "&~*",
  sibling: "&+*",
  override: "&&"
};

// src/internal/dom.ts
var STYLE_ELEMENT_ID = "__twind";
var getStyleElement = (nonce) => {
  let element = self[STYLE_ELEMENT_ID];
  if (!element) {
    element = document.head.appendChild(document.createElement("style"));
    element.id = STYLE_ELEMENT_ID;
    nonce && (element.nonce = nonce);
    element.appendChild(document.createTextNode(""));
  }
  return element;
};

// src/twind/sheets.ts
var cssomSheet = ({
  nonce,
  target = getStyleElement(nonce).sheet
} = {}) => ({
  target,
  insert: target.insertRule.bind(target)
});
var voidSheet = () => ({
  target: null,
  insert: noop
});

// src/twind/modes.ts
var mode = (report) => ({
  unknown(section, key = [], optional, context) {
    if (!optional) {
      this.report({id: "UNKNOWN_THEME_VALUE", key: join([section, ...key], ".")}, context);
    }
  },
  report({id, ...info}) {
    const message = `[${id}] ${JSON.stringify(info)}`;
    const stack = (new Error(message).stack || message).split("at ");
    for (let frame; (frame = stack.splice(1, 1)[0]) && !/(^|\.)(tw|setup) /.test(frame); ) {
    }
    report(stack.join("at "));
  }
});
var warn = mode((message) => console.warn(message));
var strict = mode((message) => {
  throw new Error(message);
});
var silent = mode(noop);

// src/twind/prefix.ts
var style_vendorizer = __toModule(require("style-vendorizer"));
var noprefix = (property2, value, important) => `${property2}:${value}${important ? " !important" : ""}`;
var autoprefix = (property2, value, important) => {
  let cssText = "";
  const propertyAlias = style_vendorizer.cssPropertyAlias(property2);
  if (propertyAlias)
    cssText += `${noprefix(propertyAlias, value, important)};`;
  let flags = style_vendorizer.cssPropertyPrefixFlags(property2);
  if (flags & 1)
    cssText += `-webkit-${noprefix(property2, value, important)};`;
  if (flags & 2)
    cssText += `-moz-${noprefix(property2, value, important)};`;
  if (flags & 4)
    cssText += `-ms-${noprefix(property2, value, important)};`;
  flags = style_vendorizer.cssValuePrefixFlags(property2, value);
  if (flags & 1)
    cssText += `${noprefix(property2, `-webkit-${value}`, important)};`;
  if (flags & 2)
    cssText += `${noprefix(property2, `-moz-${value}`, important)};`;
  if (flags & 4)
    cssText += `${noprefix(property2, `-ms-${value}`, important)};`;
  cssText += noprefix(property2, value, important);
  return cssText;
};

// src/twind/theme.ts
var ratios = (start, end) => {
  const result = {};
  do {
    for (let dividend = 1; dividend < start; dividend++) {
      result[`${dividend}/${start}`] = Number((dividend / start * 100).toFixed(6)) + "%";
    }
  } while (++start <= end);
  return result;
};
var exponential = (stop, unit, start = 0) => {
  const result = {};
  for (; start <= stop; start = start * 2 || 1) {
    result[start] = start + unit;
  }
  return result;
};
var linear = (stop, unit = "", divideBy = 1, start = 0, step = 1) => {
  const result = {};
  for (; start <= stop; start += step) {
    result[start] = start / divideBy + unit;
  }
  return result;
};
var alias2 = (section) => (theme3) => theme3(section);
var defaultTheme = {
  screens: {
    sm: "640px",
    md: "768px",
    lg: "1024px",
    xl: "1280px",
    "2xl": "1536px"
  },
  colors: {
    transparent: "transparent",
    current: "currentColor",
    black: "#000",
    white: "#fff",
    gray: {
      50: "#f9fafb",
      100: "#f3f4f6",
      200: "#e5e7eb",
      300: "#d1d5db",
      400: "#9ca3af",
      500: "#6b7280",
      600: "#4b5563",
      700: "#374151",
      800: "#1f2937",
      900: "#111827"
    },
    red: {
      50: "#fef2f2",
      100: "#fee2e2",
      200: "#fecaca",
      300: "#fca5a5",
      400: "#f87171",
      500: "#ef4444",
      600: "#dc2626",
      700: "#b91c1c",
      800: "#991b1b",
      900: "#7f1d1d"
    },
    yellow: {
      50: "#fffbeb",
      100: "#fef3c7",
      200: "#fde68a",
      300: "#fcd34d",
      400: "#fbbf24",
      500: "#f59e0b",
      600: "#d97706",
      700: "#b45309",
      800: "#92400e",
      900: "#78350f"
    },
    green: {
      50: "#ecfdf5",
      100: "#d1fae5",
      200: "#a7f3d0",
      300: "#6ee7b7",
      400: "#34d399",
      500: "#10b981",
      600: "#059669",
      700: "#047857",
      800: "#065f46",
      900: "#064e3b"
    },
    blue: {
      50: "#eff6ff",
      100: "#dbeafe",
      200: "#bfdbfe",
      300: "#93c5fd",
      400: "#60a5fa",
      500: "#3b82f6",
      600: "#2563eb",
      700: "#1d4ed8",
      800: "#1e40af",
      900: "#1e3a8a"
    },
    indigo: {
      50: "#eef2ff",
      100: "#e0e7ff",
      200: "#c7d2fe",
      300: "#a5b4fc",
      400: "#818cf8",
      500: "#6366f1",
      600: "#4f46e5",
      700: "#4338ca",
      800: "#3730a3",
      900: "#312e81"
    },
    purple: {
      50: "#f5f3ff",
      100: "#ede9fe",
      200: "#ddd6fe",
      300: "#c4b5fd",
      400: "#a78bfa",
      500: "#8b5cf6",
      600: "#7c3aed",
      700: "#6d28d9",
      800: "#5b21b6",
      900: "#4c1d95"
    },
    pink: {
      50: "#fdf2f8",
      100: "#fce7f3",
      200: "#fbcfe8",
      300: "#f9a8d4",
      400: "#f472b6",
      500: "#ec4899",
      600: "#db2777",
      700: "#be185d",
      800: "#9d174d",
      900: "#831843"
    }
  },
  spacing: {
    px: "1px",
    0: "0px",
    ...linear(4, "rem", 4, 0.5, 0.5),
    ...linear(12, "rem", 4, 5),
    14: "3.5rem",
    ...linear(64, "rem", 4, 16, 4),
    72: "18rem",
    80: "20rem",
    96: "24rem"
  },
  durations: {
    75: "75ms",
    100: "100ms",
    150: "150ms",
    200: "200ms",
    300: "300ms",
    500: "500ms",
    700: "700ms",
    1e3: "1000ms"
  },
  animation: {
    none: "none",
    spin: "spin 1s linear infinite",
    ping: "ping 1s cubic-bezier(0, 0, 0.2, 1) infinite",
    pulse: "pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite",
    bounce: "bounce 1s infinite"
  },
  backgroundColor: alias2("colors"),
  backgroundImage: {
    none: "none"
  },
  backgroundOpacity: alias2("opacity"),
  borderColor: (theme3) => ({
    ...theme3("colors"),
    DEFAULT: theme3("colors.gray.200", "currentColor")
  }),
  borderOpacity: alias2("opacity"),
  borderRadius: {
    none: "0px",
    sm: "0.125rem",
    DEFAULT: "0.25rem",
    md: "0.375rem",
    lg: "0.5rem",
    xl: "0.75rem",
    "2xl": "1rem",
    "3xl": "1.5rem",
    full: "9999px"
  },
  borderWidth: {
    DEFAULT: "1px",
    ...exponential(8, "px")
  },
  boxShadow: {
    sm: "0 1px 2px 0 rgba(0, 0, 0, 0.05)",
    DEFAULT: "0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)",
    md: "0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)",
    lg: "0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)",
    xl: "0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)",
    "2xl": "0 25px 50px -12px rgba(0, 0, 0, 0.25)",
    inner: "inset 0 2px 4px 0 rgba(0, 0, 0, 0.06)",
    none: "none"
  },
  container: {},
  divideColor: alias2("borderColor"),
  divideOpacity: alias2("borderOpacity"),
  divideWidth: alias2("borderWidth"),
  fill: {current: "currentColor"},
  flex: {
    1: "1 1 0%",
    auto: "1 1 auto",
    initial: "0 1 auto",
    none: "none"
  },
  fontFamily: {
    sans: 'ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji"'.split(","),
    serif: 'ui-serif,Georgia,Cambria,"Times New Roman",Times,serif'.split(","),
    mono: 'ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace'.split(",")
  },
  fontSize: {
    xs: ["0.75rem", "1rem"],
    sm: ["0.875rem", "1.25rem"],
    base: ["1rem", "1.5rem"],
    lg: ["1.125rem", "1.75rem"],
    xl: ["1.25rem", "1.75rem"],
    "2xl": ["1.5rem", "2rem"],
    "3xl": ["1.875rem", "2.25rem"],
    "4xl": ["2.25rem", "2.5rem"],
    "5xl": ["3rem", "1"],
    "6xl": ["3.75rem", "1"],
    "7xl": ["4.5rem", "1"],
    "8xl": ["6rem", "1"],
    "9xl": ["8rem", "1"]
  },
  fontWeight: {
    thin: "100",
    extralight: "200",
    light: "300",
    normal: "400",
    medium: "500",
    semibold: "600",
    bold: "700",
    extrabold: "800",
    black: "900"
  },
  gap: alias2("spacing"),
  gradientColorStops: alias2("colors"),
  height: (theme3) => ({
    auto: "auto",
    ...theme3("spacing"),
    ...ratios(2, 6),
    full: "100%",
    screen: "100vh"
  }),
  inset: (theme3) => ({
    auto: "auto",
    ...theme3("spacing"),
    ...ratios(2, 4),
    full: "100%"
  }),
  keyframes: {
    spin: {
      from: {
        transform: "rotate(0deg)"
      },
      to: {
        transform: "rotate(360deg)"
      }
    },
    ping: {
      "0%": {
        transform: "scale(1)",
        opacity: "1"
      },
      "75%,100%": {
        transform: "scale(2)",
        opacity: "0"
      }
    },
    pulse: {
      "0%,100%": {
        opacity: "1"
      },
      "50%": {
        opacity: ".5"
      }
    },
    bounce: {
      "0%, 100%": {
        transform: "translateY(-25%)",
        animationTimingFunction: "cubic-bezier(0.8,0,1,1)"
      },
      "50%": {
        transform: "none",
        animationTimingFunction: "cubic-bezier(0,0,0.2,1)"
      }
    }
  },
  letterSpacing: {
    tighter: "-0.05em",
    tight: "-0.025em",
    normal: "0em",
    wide: "0.025em",
    wider: "0.05em",
    widest: "0.1em"
  },
  lineHeight: {
    none: "1",
    tight: "1.25",
    snug: "1.375",
    normal: "1.5",
    relaxed: "1.625",
    loose: "2",
    ...linear(10, "rem", 4, 3)
  },
  margin: (theme3) => ({
    auto: "auto",
    ...theme3("spacing")
  }),
  maxHeight: (theme3) => ({
    ...theme3("spacing"),
    full: "100%",
    screen: "100vh"
  }),
  maxWidth: (theme3, {breakpoints}) => ({
    none: "none",
    0: "0rem",
    xs: "20rem",
    sm: "24rem",
    md: "28rem",
    lg: "32rem",
    xl: "36rem",
    "2xl": "42rem",
    "3xl": "48rem",
    "4xl": "56rem",
    "5xl": "64rem",
    "6xl": "72rem",
    "7xl": "80rem",
    full: "100%",
    min: "min-content",
    max: "max-content",
    prose: "65ch",
    ...breakpoints(theme3("screens"))
  }),
  minHeight: {
    0: "0px",
    full: "100%",
    screen: "100vh"
  },
  minWidth: {
    0: "0px",
    full: "100%",
    min: "min-content",
    max: "max-content"
  },
  opacity: {
    ...linear(100, "", 100, 0, 10),
    5: "0.05",
    25: "0.25",
    75: "0.75",
    95: "0.95"
  },
  order: {
    first: "-9999",
    last: "9999",
    none: "0",
    ...linear(12, "", 1, 1)
  },
  outline: {
    none: ["2px solid transparent", "2px"],
    white: ["2px dotted white", "2px"],
    black: ["2px dotted black", "2px"]
  },
  padding: alias2("spacing"),
  placeholderColor: alias2("colors"),
  placeholderOpacity: alias2("opacity"),
  ringColor: (theme3) => ({
    DEFAULT: theme3("colors.blue.500", "#3b82f6"),
    ...theme3("colors")
  }),
  ringOffsetColor: alias2("colors"),
  ringOffsetWidth: exponential(8, "px"),
  ringOpacity: (theme3) => ({
    DEFAULT: "0.5",
    ...theme3("opacity")
  }),
  ringWidth: {
    DEFAULT: "3px",
    ...exponential(8, "px")
  },
  rotate: {
    ...exponential(2, "deg"),
    ...exponential(12, "deg", 3),
    ...exponential(180, "deg", 45)
  },
  scale: {
    0: "0",
    50: ".5",
    75: ".75",
    ...linear(110, "", 100, 90, 5),
    125: "1.25",
    150: "1.5"
  },
  skew: {
    ...exponential(2, "deg"),
    ...exponential(12, "deg", 3)
  },
  space: alias2("spacing"),
  stroke: {
    current: "currentColor"
  },
  strokeWidth: linear(2),
  textColor: alias2("colors"),
  textOpacity: alias2("opacity"),
  transitionDuration: (theme3) => ({
    DEFAULT: "150ms",
    ...theme3("durations")
  }),
  transitionDelay: alias2("durations"),
  transitionProperty: {
    none: "none",
    all: "all",
    DEFAULT: "background-color,border-color,color,fill,stroke,opacity,box-shadow,transform",
    colors: "background-color,border-color,color,fill,stroke",
    opacity: "opacity",
    shadow: "box-shadow",
    transform: "transform"
  },
  transitionTimingFunction: {
    DEFAULT: "cubic-bezier(0.4,0,0.2,1)",
    linear: "linear",
    in: "cubic-bezier(0.4,0,1,1)",
    out: "cubic-bezier(0,0,0.2,1)",
    "in-out": "cubic-bezier(0.4,0,0.2,1)"
  },
  translate: (theme3) => ({
    ...theme3("spacing"),
    ...ratios(2, 4),
    full: "100%"
  }),
  width: (theme3) => ({
    auto: "auto",
    ...theme3("spacing"),
    ...ratios(2, 6),
    ...ratios(12, 12),
    screen: "100vw",
    full: "100%",
    min: "min-content",
    max: "max-content"
  }),
  zIndex: {
    auto: "auto",
    ...linear(50, "", 1, 0, 10)
  }
};
var flattenColorPalette = (colors, target = {}, prefix2 = []) => {
  Object.keys(colors).forEach((property2) => {
    const value = colors[property2];
    if (property2 === "DEFAULT") {
      target[join(prefix2)] = value;
      target[join(prefix2, ".")] = value;
    }
    const key = [...prefix2, property2];
    target[join(key)] = value;
    target[join(key, ".")] = value;
    if (object(value)) {
      flattenColorPalette(value, target, key);
    }
  }, target);
  return target;
};
var resolveContext = {
  negative: () => ({}),
  breakpoints: (source) => Object.keys(source).reduce((target, key) => {
    target["screen-" + key] = source[key];
    return target;
  }, {})
};
var makeThemeResolver = (config) => {
  const cache = new Map();
  const theme3 = {...defaultTheme, ...config};
  const deref = (theme4, section) => {
    const base = theme4 && theme4[section];
    const value = fn(base) ? base(resolve, resolveContext) : base;
    return value && section === "colors" ? flattenColorPalette(value) : value;
  };
  const resolve = (section, key, defaultValue) => {
    const keypath = section.split(".");
    section = keypath[0];
    if (keypath.length > 1) {
      defaultValue = key;
      key = join(tail(keypath), ".");
    }
    let base = cache.get(section);
    if (!base) {
      cache.set(section, base = {
        ...deref(theme3, section),
        ...deref(theme3.extend, section)
      });
    }
    if (key != null) {
      const value = base[(Array.isArray(key) ? join(key) : key) || "DEFAULT"];
      return value == null ? defaultValue : Array.isArray(value) && !includes(["fontSize", "outline"], section) ? join(value, ",") : value;
    }
    return base;
  };
  return resolve;
};

// src/twind/parse.ts
var groupings;
var rules;
var startGrouping = (value = "") => {
  groupings.push(value);
  return "";
};
var endGrouping = (isWhitespace) => {
  groupings.length = Math.max(groupings.lastIndexOf("") + ~~isWhitespace, 0);
};
var onlyPrefixes = (s) => s && s[0] !== ":";
var onlyVariants = (s) => s[0] === ":";
var addRule = (directive, negate, important) => {
  rules.push({
    v: groupings.filter(onlyVariants),
    d: directive,
    n: negate,
    i: important,
    $: ""
  });
};
var saveRule = (buffer) => {
  const negate = buffer[0] === "-";
  if (negate) {
    buffer = tail(buffer);
  }
  const important = buffer[buffer.length - 1] === "!";
  if (important) {
    buffer = buffer.slice(0, -1);
  }
  const prefix2 = join(groupings.filter(onlyPrefixes));
  addRule(buffer === "&" ? prefix2 : (prefix2 && prefix2 + "-") + buffer, negate, important);
  return "";
};
var parseString = (token, isVariant) => {
  let char;
  let buffer = "";
  for (let position2 = 0; position2 < token.length; ) {
    switch (char = token[position2++]) {
      case ":":
        buffer = buffer && startGrouping(":" + buffer);
        break;
      case "(":
        buffer = buffer && startGrouping(buffer);
        startGrouping();
        break;
      case ")":
      case " ":
      case "	":
      case "\n":
      case "\r":
        buffer = buffer && saveRule(buffer);
        endGrouping(char !== ")");
        break;
      default:
        buffer += char;
    }
  }
  if (buffer) {
    if (isVariant) {
      startGrouping(":" + buffer);
    } else if (buffer.slice(-1) === "-") {
      startGrouping(buffer.slice(0, -1));
    } else {
      saveRule(buffer);
    }
  }
};
var parseGroupedToken = (token) => {
  startGrouping();
  parseToken(token);
  endGrouping();
};
var parseGroup = (key, token) => {
  if (token) {
    startGrouping();
    const isVariant = includes("tbu", (typeof token)[1]);
    parseString(key, isVariant);
    if (isVariant) {
      parseGroupedToken(token);
    }
    endGrouping();
  }
};
var parseToken = (token) => {
  switch (typeof token) {
    case "string":
      parseString(token);
      break;
    case "function":
      addRule(token);
      break;
    case "object":
      if (Array.isArray(token)) {
        token.forEach(parseGroupedToken);
      } else if (token) {
        Object.keys(token).forEach((key) => {
          parseGroup(key, token[key]);
        });
      }
  }
};
var staticsCaches = new WeakMap();
var buildStatics = (strings) => {
  let statics = staticsCaches.get(strings);
  if (!statics) {
    statics = strings.map((token) => {
      const staticRules = rules = [];
      parseString(token);
      const activeGroupings = [...groupings];
      rules = [];
      return (interpolation) => {
        rules.push(...staticRules);
        groupings = [...activeGroupings];
        interpolation && parseToken(interpolation);
      };
    });
    staticsCaches.set(strings, statics);
  }
  return statics;
};
var parse = (tokens) => {
  groupings = [];
  rules = [];
  if (Array.isArray(tokens[0]) && Array.isArray(tokens[0].raw)) {
    buildStatics(tokens[0]).forEach((apply, index) => apply(tokens[index + 1]));
  } else {
    parseToken(tokens);
  }
  return rules;
};

// src/twind/translate.ts
var translate = (plugins2, context) => (rule) => {
  if (fn(rule.d)) {
    return rule.d(context);
  }
  const parameters = rule.d.split("-");
  if (parameters[0] === "tw" && rule.$ === rule.d) {
    return rule.$;
  }
  for (let index = parameters.length; index; index--) {
    const id = join(parameters.slice(0, index));
    const plugin = plugins2[id];
    if (plugin) {
      return fn(plugin) ? plugin(tail(parameters, index), context, id) : string(plugin) ? context.tw(plugin) : plugin;
    }
  }
};

// src/twind/decorate.ts
var decorate = (darkMode, variants2, {theme: theme3, tag}) => {
  const applyVariant = (translation, variant) => {
    const size = theme3("screens", tail(variant), "");
    if (size) {
      return {[`@media (min-width: ${size})`]: translation};
    }
    if (variant === ":dark" && darkMode === "class") {
      return {[`.dark &`]: translation};
    }
    if (variant.slice(1, 7) === "group-") {
      return {[`.${escape(tag("group"))}:${tail(variant, 7)} &`]: translation};
    }
    return {[variants2[tail(variant)] || "&" + variant]: translation};
  };
  return (translation, rule) => rule.v.reduceRight(applyVariant, translation);
};

// src/twind/presedence.ts
var _2;
var responsivePrecedence = (css) => (((_2 = /(?:^|min-width: *)(\d+(?:.\d+)?)(p)?/.exec(css)) ? +_2[1] / (_2[2] ? 15 : 1) / 10 : 0) & 31) << 23;
var seperatorPrecedence = (string2) => {
  _2 = 0;
  for (let index = string2.length; index--; ) {
    _2 += includes("-:,", string2[index]);
  }
  return _2;
};
var atRulePresedence = (css) => (seperatorPrecedence(css) & 15) << 18;
var PRECEDENCES_BY_PSEUDO_CLASS = [
  "rst",
  "st",
  "en",
  "d",
  "nk",
  "sited",
  "pty",
  "ecked",
  "cus-w",
  "ver",
  "cus",
  "cus-v",
  "tive",
  "sable",
  "ad-on",
  "tiona",
  "quire"
];
var pseudoPrecedence = (pseudoClass) => 1 << (~(_2 = PRECEDENCES_BY_PSEUDO_CLASS.indexOf(pseudoClass.replace(/^:group-/, ":").slice(3, 8))) ? _2 : 17);
var makeVariantPresedenceCalculator = (theme3, variants2) => (presedence2, variant) => presedence2 | ((_2 = theme3("screens", tail(variant), "")) ? responsivePrecedence(_2) : variant === ":dark" ? 1 << 22 : (_2 = variants2[variant] || variant)[0] === "@" ? atRulePresedence(_2) : pseudoPrecedence(variant));
var declarationPropertyPrecedence = (property2) => property2[0] === "-" ? 0 : seperatorPrecedence(property2) + ((_2 = /^(?:(border-(?!w|c|sty)|[tlbr].{2,4}m?$|c.{7}$)|([fl].{5}l|g.{8}$|pl))/.exec(property2)) ? +!!_2[1] || -!!_2[2] : 0) + 1;

// src/twind/serialize.ts
var stringifyBlock = (body, selector) => selector + "{" + body + "}";
var serialize = (prefix2, variants2, context) => {
  const {theme: theme3, tag} = context;
  const tagVar = (_3, property2) => "--" + tag(property2);
  const tagVars = (value) => `${value}`.replace(/--(tw-[\w-]+)\b/g, tagVar);
  const stringifyDeclaration = (property2, value, important) => {
    property2 = tagVars(property2);
    return Array.isArray(value) ? join(value.filter(Boolean).map((value2) => prefix2(property2, tagVars(value2), important)), ";") : prefix2(property2, tagVars(value), important);
  };
  let rules2;
  const stringify = (atRules, selector, presedence2, css, important) => {
    let declarations = "";
    let maxPropertyPresedence = 0;
    let numberOfDeclarations = 0;
    Object.keys(css).forEach((key) => {
      let value = css[key];
      if (fn(value)) {
        value = value(context);
      }
      if (includes("rg", (typeof value)[5]) || Array.isArray(value)) {
        const property2 = hyphenate(key);
        numberOfDeclarations += 1;
        maxPropertyPresedence = Math.max(maxPropertyPresedence, declarationPropertyPrecedence(property2));
        declarations = (declarations && declarations + ";") + stringifyDeclaration(property2, value, important);
      } else if (value) {
        if (key[0] === "@") {
          if (key[1] === "f") {
            stringify([], key, 0, value, important);
          } else if (key[1] === "k") {
            const currentSize = rules2.length;
            stringify([], "", 0, value, important);
            const waypoints = rules2.splice(currentSize, rules2.length - currentSize);
            rules2.push({
              r: stringifyBlock(join(waypoints.map((p) => p.r), ""), key),
              p: waypoints.reduce((sum, p) => sum + p.p, 0)
            });
          } else {
            stringify([...atRules, key], selector, presedence2 | (responsivePrecedence(key) || atRulePresedence(key)), value, important);
          }
        } else {
          stringify(atRules, selector ? selector.replace(/([^,])+/g, (selectorPart) => key.replace(/([^,])+/g, (keyPart) => includes(keyPart, "&") ? keyPart.replace(/&/g, selectorPart) : (selectorPart && selectorPart + " ") + keyPart)) : key, selector && includes(key, "&") ? presedence2 : 0, value, important);
        }
      }
    });
    if (numberOfDeclarations) {
      rules2.push({
        r: atRules.reduceRight(stringifyBlock, stringifyBlock(declarations, selector)),
        p: presedence2 * (1 << 8) + (((maxPropertyPresedence || 15) & 15) << 4 | Math.max(0, 15 - numberOfDeclarations) & 15)
      });
    }
  };
  const variantPresedence = makeVariantPresedenceCalculator(theme3, variants2);
  return (css, className, rule) => {
    rules2 = [];
    stringify([], className ? "." + escape(className) : "", rule ? rule.v.reduceRight(variantPresedence, 0) : 0, css, rule && rule.i);
    return rules2;
  };
};

// src/twind/inject.ts
var inject = (sheet, mode2, init, context) => {
  let sortedPrecedences;
  init((value = []) => sortedPrecedences = value);
  let insertedRules;
  init((value = new Set()) => insertedRules = value);
  return ({r: css, p: presedence2}) => {
    if (!insertedRules.has(css)) {
      insertedRules.add(css);
      const index = sortedInsertionIndex(sortedPrecedences, presedence2);
      try {
        sheet.insert(css, index);
        sortedPrecedences.splice(index, 0, presedence2);
      } catch (error) {
        if (!/:-[mwo]/.test(css)) {
          mode2.report({id: "INJECT_CSS_ERROR", css, error}, context);
        }
      }
    }
  };
};

// src/twind/configure.ts
var sanitize = (value, defaultValue, disabled, enabled = defaultValue) => value === false ? disabled : value === true ? enabled : value || defaultValue;
var loadMode = (mode2) => (string(mode2) ? {t: strict, a: warn, i: silent}[mode2[1]] : mode2) || warn;
var toString = (rule, directive = rule.d) => {
  if (fn(directive))
    return "";
  const base = join(rule.v, "");
  return (base && tail(base) + ":") + (rule.n ? "-" : "") + directive + (rule.i ? "!" : "");
};
var configure = (config = {}) => {
  const theme3 = makeThemeResolver(config.theme);
  const mode2 = loadMode(config.mode);
  const hash = sanitize(config.hash, false, false, cyrb32);
  let activeRule = {v: []};
  const context = {
    tw: (...tokens) => process(tokens),
    theme: (section, key, defaultValue) => {
      if (key != null && !key.length) {
        key = "DEFAULT";
      }
      const value = theme3(section, key, defaultValue) || mode2.unknown(section, key == null || Array.isArray(key) ? key : key.split("."), defaultValue != null, context);
      return activeRule.n && value && string(value) ? `calc(${value} * -1)` : value;
    },
    tag: (value) => hash ? hash(value) : value
  };
  const translate3 = translate({...corePlugins, ...config.plugins}, context);
  const tryTranslate = (rule) => {
    const parentRule = activeRule;
    activeRule = rule;
    try {
      return translate3(rule);
    } finally {
      activeRule = parentRule;
    }
  };
  const variants2 = {...coreVariants, ...config.variants};
  const decorate3 = decorate(config.darkMode || "media", variants2, context);
  const serialize3 = serialize(sanitize(config.prefix, autoprefix, noprefix), variants2, context);
  const sheet = config.sheet || (typeof window === "undefined" ? voidSheet() : cssomSheet(config));
  const {init = (callback) => callback()} = sheet;
  const inject3 = inject(sheet, mode2, init, context);
  let idToClassName;
  init((value = new Map()) => idToClassName = value);
  const inlineDirectiveName = new WeakMap();
  const evaluateFunctions = (key, value) => fn(value) ? value(context) : value;
  const convert = (rule) => {
    if (activeRule.v.length) {
      rule = {...rule, v: [...activeRule.v, ...rule.v], $: ""};
    }
    if (!rule.$) {
      rule.$ = toString(rule) || toString(rule, inlineDirectiveName.get(rule.d));
    }
    let className = idToClassName.get(rule.$);
    if (className == null) {
      let translation = tryTranslate(rule);
      if (!rule.$) {
        rule.$ = cyrb32(JSON.stringify(translation, evaluateFunctions));
        inlineDirectiveName.set(rule.d, rule.$);
        rule.$ = toString(rule, rule.$);
      }
      if (string(translation)) {
        className = translation;
      } else if (object(translation)) {
        translation = decorate3(translation, rule);
        className = hash ? hash(JSON.stringify(translation, evaluateFunctions)) : rule.$;
        serialize3(translation, className, rule).forEach(inject3);
      } else {
        className = rule.$;
        mode2.report({id: "UNKNOWN_DIRECTIVE", rule: className}, context);
      }
      idToClassName.set(rule.$, className);
      if (idToClassName.size > 3e4) {
        idToClassName.delete(idToClassName.keys().next().value);
      }
    }
    return className;
  };
  const process = (tokens) => parse(tokens).map(convert).filter(Boolean).join(" ");
  const preflight2 = sanitize(config.preflight, identity, false);
  if (preflight2) {
    const css = createPreflight(theme3);
    const styles = serialize3(fn(preflight2) ? preflight2(css, context) || css : {...css, ...preflight2});
    init((injected = (styles.forEach(inject3), true)) => injected);
  }
  return {
    init: () => mode2.report({id: "LATE_SETUP_CALL"}, context),
    process,
    theme: theme3
  };
};

// src/twind/instance.ts
var create = (config) => {
  let process = (tokens) => {
    init();
    return process(tokens);
  };
  let theme3 = (section, key, defaultValue) => {
    init();
    return theme3(section, key, defaultValue);
  };
  let init = (config2) => {
    ;
    ({process, init, theme: theme3} = configure(config2));
  };
  if (config)
    init(config);
  return {
    tw: (...tokens) => process(tokens),
    setup: (config2) => init(config2),
    theme: theme3
  };
};

// src/index.ts
var instance2 = create();
var {tw} = instance2;
var {setup} = instance2;
var {theme: theme2} = instance2;
